src/services/pcn-packetcapture/src/Packetcapture.cpp

```diff
@@ -0,0 +1,358 @@
+  filters_table ft_init;
+  ft_init.dst_port_flag = ft_init.src_port_flag = ft_init.network_filter_src_flag = ft_init.network_filter_dst_flag = ft_init.l4proto_flag = false;
+  t_filters_in.set(0x0, ft_init);
+  t_filters_out.set(0x0, ft_init);
+  writeHeader = true;
+  dt = "";
+}
+
+void Packetcapture::writeDump(const std::vector<uint8_t> &packet){
+  PacketJsonObject pj;
+  auto p = std::shared_ptr<Packet>(new Packet(*this, pj));
+  struct timeval tp;
+  std::streamsize len;
+
+  if( dt == "" ){
+    time_t now = time(0);
+    tm *ltm = localtime(&now);
+    std::stringstream tmstmp;
+    tmstmp << ltm->tm_mday << "-" << ltm->tm_mon << "-" << (ltm->tm_year + 1900) << "-" << ltm->tm_hour << ":" << ltm->tm_min << ":" << ltm->tm_sec;
+    dt = tmstmp.str();
+  }
+  
+  myFile.open("capture_"+ dt +".pcap", std::ios::binary | std::ios::app);
+
+  if( writeHeader == true ){
+    struct pcap_file_header *pcap_header = new struct pcap_file_header;
+    pcap_header->magic = global_header->getMagic();
+    pcap_header->version_major = global_header->getVersionMajor();
+    pcap_header->version_minor = global_header->getVersionMinor();
+    pcap_header->thiszone = global_header->getThiszone();   //timestamp are always in GMT
+    pcap_header->sigfigs = global_header->getSigfigs();
+    pcap_header->snaplen = filters->getSnaplen();
+    pcap_header->linktype = global_header->getLinktype();
+
+    myFile.write(reinterpret_cast<const char*>(pcap_header), sizeof(*pcap_header));
+    writeHeader = false;
+  }
+  gettimeofday(&tp, NULL);
+  p->setTimestampSeconds((uint32_t) tp.tv_sec);
+  p->setTimestampMicroseconds((uint32_t) tp.tv_usec);
+  p->setPacketlen((uint32_t) packet.size());
+  if(p->getPacketlen() > filters->getSnaplen()){
+    p->setCapturelen(filters->getSnaplen());
+  }else{
+    p->setCapturelen(p->getPacketlen());
+  }
+  p->setRawPacketData(packet);
+
+  struct pcap_pkthdr *pkt_hdr = new struct pcap_pkthdr;
+  pkt_hdr->ts_sec = p->getTimestampSeconds();
+  pkt_hdr->ts_usec = p->getTimestampMicroseconds();
+  pkt_hdr->len = p->getPacketlen();
+  pkt_hdr->caplen = p->getCapturelen();
+  filters->getSnaplen() < p->getRawPacketData().size() ? len = filters->getSnaplen() : len = p->getRawPacketData().size();
+
+  myFile.write(reinterpret_cast<const char*>(pkt_hdr), sizeof(*pkt_hdr));
+  myFile.write(reinterpret_cast<const char*>(&p->getRawPacketData()[0]), len );
+  myFile.close();
+}
+
+Packetcapture::~Packetcapture() {
+  logger()->info("Destroying Packetcapture instance");
+}
+
+void Packetcapture::packet_in(polycube::service::Direction direction,
+    polycube::service::PacketInMetadata &md,
+    const std::vector<uint8_t> &packet) {
+  
+  Tins::EthernetII pkt(&packet[0], packet.size());
+      
+  switch (direction) {
+    case polycube::service::Direction::INGRESS:
+    if( getNetworkmode() == true ){
+      addPacket(packet);    /* store the packet */
```

Comment:
"store the packet" means "Store packet in the FIFO buffer", correct?

Reply:
That is correct, afther this instruction the packet will be stored in the FIFO queue